Refactor builder with new Go interfaces

tiborvass opened the pull request:
This patch creates interfaces in builder/ for building Docker images.
It is a first step in a series of patches to remove the daemon
dependency on builder and later allow a client-side Dockerfile builder
as well as potential builder plugins (aka builders other than from Dockerfiles)

builder/dockerfile/ and api/server/builder.go contain implementations
of those interfaces as a refactoring of the current code.

Signed-off-by: Tibor Vass tibor@docker.com

Ping @duglin @jlhawn @vdemeester 

TODO (must have):
- [x] Cancelling a build
- [x] Restore safe rmi when pulling (#13681)

TODO (good to have):
- [x] Restoring progress bar?
- [x] Handle warnings

Commit:
Move builder files to builder/dockerfile

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
Abstract builder and implement server-side dockerfile builder

This patch creates interfaces in builder/ for building Docker images.
It is a first step in a series of patches to remove the daemon
dependency on builder and later allow a client-side Dockerfile builder
as well as potential builder plugins.

It is needed because we cannot remove the /build API endpoint, so we
need to keep the server-side Dockerfile builder, but we also want to
reuse the same Dockerfile parser and evaluator for both server-side and
client-side.

builder/dockerfile/ and api/server/builder.go contain implementations
of those interfaces as a refactoring of the current code.

Signed-off-by: Tibor Vass <tibor@docker.com>

vdemeester commented:
ah you got kick real quick by `golint` :sweat_smile:

stevvooe commented on builder/builder.go:
```diff
@@ -0,0 +1,77 @@
+
+type Context interface {
+	io.Closer
+	Get(path string) (ContextEntry, error)
```
I would just call this `Stat(path string) (os.FileInfo, error)` and allow interface conversion to `Pather`, with the following definition:

```
type Pather interface {
    Path() string
}
```

tiborvass replied:
Yeah was thinking about this too, feel free to give more input and I'll update accordingly.

stevvooe commented on builder/builder.go:
```diff
@@ -0,0 +1,77 @@
+type ContextEntry interface {
+	os.FileInfo
+	Path() string
+	Open() (io.ReadCloser, error)
```
I'd recommend decoupling the opening of resources from the description of a resource. Having `Open` here would be similar to have an `Open` method on URL. This should only be called through the context. Removing this here ensures resources can only be opened with a single code path.

tiborvass replied:
I can do this too, I believe I added it for convenience but I'll look into removing it.

stevvooe replied:
@tiborvass I would recommend trying to remove this interface and work directly with `os.FileInfo` and see if you can get it working. My "jury" is still out on whether an `os.FileInfo` should be trusted for more than just its basename but that is the model the stdlib follows and is common to unix systems. This likely comes from the idea that the path is actually contextual, rather than a property of the file system entry itself. For example, the path of an `os.FileInfo` may be completely differently from one context to the next but could have the same `Name()` and inode. Albeit, this consideration may not be valid for this use case, but may help to clean up the data model.

Let me know what your experience here is, as I have a similar problem in continuity.

stevvooe commented on builder/builder.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func (fi FileInfo) Open() (io.ReadCloser, error) {
+	return os.Open(fi.FilePath)
```
Having this as `os.Open`, rather than going through the context breaks the interface of `Context`. I'd recommend removing the requirement for the `Open` method on `ContextEntry`.

tiborvass replied:
Agreed, like said in the earlier comment

stevvooe commented on builder/builder.go:
```diff
@@ -0,0 +1,77 @@
+	return os.Open(fi.FilePath)
+}
+
+func (fi FileInfo) Stat() os.FileInfo {
```
This is kind of odd. `os.FileInfo` already has a `Sys()` method to support this "unpacking" use case.

tiborvass replied:
I removed Stat() already I'll push as soon as I'm done golinting.
If I remove Open() too, I can remove FileInfo altogether and just keep HashFileInfo.
Also, I used to use `Sys()` but I changed it to just casting the interface to `(*HashFileInfo)` to check if it supports hashes. I can change it back to using `Sys()` if you prefer

stevvooe replied:
@tiborvass I'd leave `Sys()` to return the actual `syscall.Stat_t` and then just have interface extensions that can be type-asserted:

```
switch v := fi.(type) {
case Hashed:
    h := v.Hash()
default:
    rd, err := ctx.Open(p)
    if err != nil {
        return err
    }

    io.Copy(digester, rd)
}
```

stevvooe commented on builder/builder.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+type ImageCache interface {
+	GetCachedImage(ImageID, *runconfig.Config) (ImageID, error)
```
Why does a caching method take the configuration? Do we not trust the image id?

crosbymichael commented:
@tiborvass is there  a specific commit we should be reviewing or is there something specific we should look at since alot of the code changes were just moving existing code around? thx!

tiborvass commented:
@stevvooe I updated with your changes, I'm not sure I'm convinced because it makes the code at some places not type-safe anymore, and has substantial redundant runtime type-checking.

@crosbymichael yeah I tried to separate the move from builder to builder/dockerfile/ in the first commit and do all the refactoring in the second commit, but refactoring is inevitably moving code around so I apologize in advance. When in doubt just ask, I can tell you where things come from.

stevvooe commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,688 @@
+			builder.Hashed
+		})
+		if !ok {
+			return fmt.Errorf("builder: unexpected type (%T) for FileInfo", info)
```
This is a case where I would change the type signature of the `WalkFunc` to require a `Hashed`. I would not require `Pather` here, since we get it as a method argument.

tiborvass replied:
@stevvooe I can't require a Hashed, because not all contexts will necessarily have support for caching.

stevvooe commented on api/server/builder.go:
```diff
@@ -0,0 +1,194 @@
+	return d.Daemon.Commit(c, cfg)
+}
+
+func (d builderDocker) Copy(c *daemon.Container, destPath string, srcInfo os.FileInfo, decompress bool) (err error) {
```
Here is another case where I would avoid carrying the path information with the `os.FileInfo` object, removing the need for the `Pather` cast. I would simply pass in the path for the source, in addition to the file info.

stevvooe commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,688 @@
+			if !allowRemote {
+				return fmt.Errorf("Source can't be a URL for %s", cmdName)
+			}
+			fi, err = b.download(orig)
```
Just return the path here, in addition to the info about the file.

LK4D4 replied:
Forgotten comment?

tiborvass replied:
@LK4D4 no. I addressed it in a previous version and had to revert because it complicates things a lot. I can add a `TODO` for later fix. It should get simplified once `builder.PathFileInfo` etc. and all those are removed.

stevvooe commented:
@tiborvass I see the concerns in the code. I've put some comments in-line. I really believe separating path from the other attributes is adequate, so `Pather` may not be needed. In some places, where only `Hashed` is used, I'd recommend taking that as the type to the function.

jfrazelle commented:
hmmm somme tests failed wrt build

tiborvass commented:
Now that the buildarg PR is merged, I rebased this PR. Also all tests should now be passing.

The only thing that's changing UX-wise, is that I removed the 2 progressbars (the code is commented out on purpose so that people can see the issue and maybe help). I still find the issue minor. Let me know what you think.

Ping @crosbymichael @jfrazelle @duglin

duglin commented on api/client/build.go:
```diff
@@ -131,10 +130,19 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 
-	excludes, err := utils.ReadDockerIgnore(path.Join(contextDir, ".dockerignore"))
-	if err != nil {
+	f, err := os.Open(filepath.Join(contextDir, ".dockerignore"))
```
Was this change made so that you could pass in a reader instead of a path?  If so, why?

tiborvass replied:
Because on the server side, I only have a reader when reading from a context: https://github.com/docker/docker/pull/16147/files#diff-ca102c238cfd962b0905c8009ad2c39bR42

duglin replied:
interesting. Can you elaborate more on the future use of the Context interface? Are you planning on supporting a build context that isn't actually on disk?  Just trying to understand when we wouldn't have a real file to point to.

tiborvass replied:
You are right that it is not strictly needed for a client-side builder, however if in the future we want to support builder plugins, it makes sense to abstract the context to allow for more possibilities. I'm also fine changing the Context interface to just handle paths instead of readers, it just seemed more appropriate like this after talking with others.

tiborvass commented on api/server/builder.go:
```diff
@@ -0,0 +1,219 @@
+
+func (d builderDocker) LookupImage(name string) (*image.Image, error) {
+	img, err := d.Daemon.Repositories().LookupImage(name)
+	logrus.Debugf("lookupimage: %s: %v", name, err)
```
leftover debug, will remove

duglin commented:
@tiborvass before I review more, I'd like to understand the bigger picture here because that will influence how I view the changes.

When we talk about a client-side builder, I've been assuming that (at a high level) we would convert each Dockerfile command into a one or more `docker xxx ...` commands (or API calls).  Which means, on the daemon side, the build funcs could all go away and we 'd just need to make sure we had all of the needed functionality exposed via APIs and (eventually) `docker xxx` CLI commands so that people could use their own scripting language.

Then the "cli builder" becomes nothing more than a translator, from Dockerfiles to API calls.

Additionally, during this change, we would "fix" all of the non-intuitive behavior that people have complained about.  For example, I'm pretty sure we'd want a new (cleaner) parser and cmd syntax that is consistent across all commands. The number of commands we have might change (e.g. drop either COPY or ADD).

So, with all of that, does it really make sense to morph the current build stuff at all?  In other words, should we start over because the design is so radically different? And use this opportunity to break free from the old pains. The CLI builder would be v2, the current stuff is v1, but deprecated.

But I'd like to understand what your vision is for this.

/cc @jlhawn

tiborvass commented:
@duglin thanks for asking me to clarify, I forgot to put this PR (pun alert) in its context.

The main reason why we need this PR instead of just redoing a client-side builder PR is that we still need the server-side builder, we can't just have `/build` not work anymore. The other reason is that we already have a Dockerfile parser and evaluator, however baroque it is, that's what it currently is and we don't want to break people. So we decided we wanted to reuse the same Dockerfile-specific code for both the server side and the client side, which meant: decoupling daemon from the builder code.

The secondary objective of this PR is to make the Dockerfile code, ONE implementation (the only for now) for the builder. Anything that's not dockerfile specific should be in builder/.

You are correct in that the client-side builder is nothing more than translating Dockerfile commands into API calls. The API calls will be what the builder.Docker interface will become. The server-side implementation just calls out to the Daemon object directly.

After this PR I will send the second part, that gets rid of the daemon dependency altogether and handles the commit `--change` codepath as well.

LK4D4 commented on api/client/build.go:
```diff
@@ -131,10 +130,19 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
-	excludes, err := utils.ReadDockerIgnore(path.Join(contextDir, ".dockerignore"))
-	if err != nil {
+	f, err := os.Open(filepath.Join(contextDir, ".dockerignore"))
+	switch {
```
It should be just like `if err != nil && !os.IsNotExist(err) { return err }`

tiborvass replied:
thanks will fix, can't remember why i used a switch here, maybe the code changed.

LK4D4 commented on api/server/builder.go:
```diff
@@ -0,0 +1,219 @@
+)
+
+type builderDocker struct {
+	*daemon.Daemon
```
As I see you always using `d.Daemon`. Why then embed field? Maybe just `Daemon: *daemon.Daemon`?

tiborvass replied:
you're correct, I think I embedded it thinking it would be a smaller diff, but you're right I'll change it.

LK4D4 commented:
Oh, wait, it's design-review, sorry :)

duglin commented:
I wasn't suggesting that we drop the `/build` APIs right away, and we can even keep them forever I guess.  But, I think it would be a mistake to introduce a new `docker build` type of command (and I do think we'd need a new one so people can choose whether they want cli vs daemon builds) and have it keep all of the things about the current command that people don't like. The minute we force people to choose which command to use we have to explain why we have two ways to do the same thing. So, I'd prefer if we didn't duplicate features, rather we use this opportunity to version the builder and make the CLI-builder into our v2 and fix things.

Additionally, as your PR is showing, its a lot of work to head down the path you're going, and we're probably going to end up adding new APIs **just** to support this - meaning we'll have new APIs for no other reason other than to support semantics that make no sense except in the context of the current build design. I suspect that if we were to start from scratch we wouldn't have those APIs, and that makes me uncomfortable.

Any chance of revisiting this decision and head down a path that would benefit our users more?

tiborvass commented:
@duglin i'm not sure I follow what you're saying about two docker commands. There won't be two docker commands, I intend to reuse `docker build` to be client-side (in a future PR). Dockerfile v2 is a completely orthogonal issue that is a lot more complicated. We need client-side builder for trust to work well too. UX issues are always more complicated. This is simply refactoring the existing code, plus we always wanted to make builder more standalone so this is also going in that direction.

icecrime commented:
> Then the "cli builder" becomes nothing more than a translator, from Dockerfiles to API calls.

That's exactly it from my point of view too.

On the path forward, `/build` will need to exist for backward compatibility reasons. What I like about this PR, is that is shows progress in isolating the builder component, and makes interfaces emerge about what a builder _is_. This will make it easy to provide different implementations, from "local" (customizable, maybe even scriptable, ...) to "remote" (backward compatible). Ultimately the client-side one and (legacy) server-side one might be instantiated in very different ways, but will expose the same responsibility behind the same interface.

jfrazelle commented:
design LGTM

tonistiigi commented:
design LGTM

swernli commented:
I verified that the merged branch works with Windows long paths, preserving the fix merged in https://github.com/docker/docker/pull/15898.

jfrazelle commented:
needs rebase but overall LGTM

jfrazelle commented:
hopefully its not a bad rebase :/

vdemeester commented on builder/builder.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// PathFileInfo is a convenience struct that implements the FileInfo interface.
+type PathFileInfo struct {
```
Same as up right, should go "once pkg/archive exposes a walk function that Context can use", if I follow correctly :sweat_smile:.

tiborvass replied:
right, but once you remove FileInfo, the rest goes with it :) I could add a comment you're right.

vdemeester commented on builder/builder.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+// Docker abstracts calls to a Docker Daemon.
+type Docker interface {
```
I think #16638 tries to do something similar (or at least *compatible*), just a note that it could be "in the end" joined (kinda :dancer:).

tiborvass replied:
It is similar indeed and I agree we should coordinate. This interface is far from final, for one thing I need to get rid of the daemon dependency. I will take a look at that PR and give him feedback, but I still think we should consider this PR as-is.

vdemeester commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+	runConfig        *runconfig.Config // runconfig for cmd, run, entrypoint etc.
+	flags            *BFlags
+	tmpContainers    map[string]struct{}
+	image            string // imageID
```
Hum, if it really is `imageID`, why not calling it like that ? (except for backward-compatibility :stuck_out_tongue:)

tiborvass replied:
I believe i just moved the code. Also the ImageID type is temporary, I could have very well just used a string directly but I preferred to type it out because we will use a reference package that deals with image references.

vdemeester commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+			return "", err
+		}
+		shortImgID = stringid.TruncateID(b.image)
+		fmt.Fprintf(b.Stdout, " ---> %s\n", shortImgID)
```
This probably is a nit (and maybe not the right PR to do that) but why not a `b.Logf(…)` or something like that ?

tiborvass replied:
Just because that's how it was before except it was called `b.Out` :) Feel free to refactor more after this PR lands :)

tiborvass commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,658 @@
+	}
+
+	// Download and dump result to tmp file
+	// TODO: put a better progressreader back
```
if someone has an idea on how to add the progress bar back... Otherwise I'll have to remove this dead code.

cpuguy83 replied:
```diff
diff --git a/builder/dockerfile/internals.go b/builder/dockerfile/internals.go
index 2319cd6..850da09 100644
--- a/builder/dockerfile/internals.go
+++ b/builder/dockerfile/internals.go
@@ -29,6 +29,8 @@ import (
 	"github.com/docker/docker/pkg/httputils"
 	"github.com/docker/docker/pkg/ioutils"
 	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/progressreader"
+	"github.com/docker/docker/pkg/streamformatter"
 	"github.com/docker/docker/pkg/stringid"
 	"github.com/docker/docker/pkg/stringutils"
 	"github.com/docker/docker/pkg/system"
@@ -271,20 +273,19 @@ func (b *Builder) download(srcURL string) (fi builder.FileInfo, err error) {
 
 	// Download and dump result to tmp file
 	// TODO: put a better progressreader back
-	/*
-		if _, err := io.Copy(tmpFile, progressreader.New(progressreader.Config{
-			      In:       resp.Body,
-			      Out:      b.Stdout,
-			      Formatter: ...,
-			      Size:     resp.ContentLength,
-			      NewLines: true,
-			      ID:       "",
-			      Action:   "Downloading",
-	*/
-	if _, err = io.Copy(tmpFile, resp.Body); err != nil {
+	if _, err = io.Copy(tmpFile, progressreader.New(progressreader.Config{
+		In:        resp.Body,
+		Out:       b.Stdout,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      resp.ContentLength,
+		NewLines:  true,
+		ID:        "",
+		Action:    "Downloading",
+	})); err != nil {
 		tmpFile.Close()
 		return
 	}
+
 	fmt.Fprintln(b.Stdout)
 	// ignoring error because the file was already opened successfully
 	tmpFileSt, _ := tmpFile.Stat()
```

cpuguy83 replied:
Progress reader looks fine to me.

tiborvass replied:
That's weird, Stdout is already streamformatted, so it seems like it should double-streamformat it...

cpuguy83 replied:
https://asciinema.org/a/ahwcg5zcknsy57i15uaialtaz

cpuguy83 replied:
Could also do this nastiness:
```
		Out:       b.Stdout.(*streamformatter.StdoutFormatter).Writer,
		Formatter: b.Stdout.(*streamformatter.StdoutFormatter).StreamFormatter,
```
Or store b.Stdout/b.Stderr as Stdout/Stderr streams respectively rather than straight up io.Writers.

cpuguy83 replied:
I think the double stream format just adds an extra `\r`.

tiborvass replied:
@cpuguy83 Sorry forgot to answer to you about this. I updated the code to use the casting for now. This whole progress bar thing complicates the code I think.

tiborvass commented on builder/remote.go:
```diff
@@ -0,0 +1,121 @@
+	// Pass through - this is a pre-packaged context, presumably
+	// with a Dockerfile with the right name inside it.
+	return MakeTarSumContext(contextReader)
+	/* TODO: progress bar
```
Same question as in https://github.com/docker/docker/pull/16147/files#r40967912

vdemeester commented on builder/tarsum.go:
```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		err = convertPathError(err, path)
+	}
+	return
```
This is a nit, but I tend to find named returns (and thus an empty `return`) confusing and hard to follow (but.. this function is small so..).. just a rant :stuck_out_tongue_closed_eyes:.

vdemeester commented on builder/git.go:
```diff
@@ -0,0 +1,28 @@
+	}
+
+	defer func() {
+		// TODO: print errors?
```
Yeah, should definitely do something there.. if failing here, it's *silent* :sweat:.

tiborvass replied:
the reason i'm asking is because it was like that in the original: https://github.com/docker/docker/pull/16147/files#diff-ff92bccd20218a5813b0004f2fba2ce0L142

vdemeester commented:
Needs some test fixing (and a little intimidating in term of review size) but overall LGTM :wink:

LK4D4 commented on api/client/build.go:
```diff
@@ -131,11 +130,19 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+	f, err := os.Open(filepath.Join(contextDir, ".dockerignore"))
+	if err != nil && !os.IsNotExist(err) {
 		return err
+	} else if err == nil {
```
I dunno why golint is so calm, but there shouldn't be else :)

LK4D4 commented on api/client/build.go:
```diff
@@ -131,11 +130,19 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	}
 
-	var includes = []string{"."}
+	var (
```
I personally hate `var` blocks :) variables should be defined where it need, includes can be not needed at all later because of error.
I noticed this because of block, but I think it's good time to move includes below to its usage

tiborvass replied:
i can do `:=`, do you prefer that?

LK4D4 replied:
I'm okay with `var` words, just includes should be defined below where it used first time.

LK4D4 commented on api/server/router/local/image.go:
```diff
@@ -56,13 +61,18 @@ func (s *router) postCommit(ctx context.Context, w http.ResponseWriter, r *http.
 	}
 
-	imgID, err := builder.Commit(cname, s.daemon, commitCfg)
+	container, err := s.daemon.Get(cname)
```
Using `Get` here is sorta violate our "weak" agreement about separation API from daemon. Maybe we can move it to `Commit`, so API won't be dependent on `Container` struct?

tiborvass replied:
@LK4D4 the API is still dependent on daemon unfortunately. I think it's still valuable to have this first PR and have another one remove completely daemon dependency. It's already a pretty big PR :S

LK4D4 commented on builder/builder.go:
```diff
@@ -0,0 +1,143 @@
+
+// ImageID is a temporary type for an Image ID.
+// It should be replaced with an image reference.
+type ImageID string
```
I didn't understand type and comment :/

tiborvass replied:
@LK4D4 I can remove the type if you prefer and just use `string` directly. I put the type to make sure we replace it later with a type from image package or a reference package (from distribution). However you want.

LK4D4 replied:
Yup, seems like patch will be simpler without it.

LK4D4 commented on api/server/router/local/image.go:
```diff
@@ -269,7 +279,7 @@ func (s *router) postBuild(ctx context.Context, w http.ResponseWriter, r *http.R
 		authConfigs        = map[string]cliconfig.AuthConfig{}
 		authConfigsEncoded = r.Header.Get("X-Registry-Config")
-		buildConfig        = builder.NewBuildConfig()
+		buildConfig        = new(dockerfile.Config)
```
same here, it should be defined way below as `buildConfig := &dockerfile.Config{}`. I'm okay with `new`, but seems like in other code we use `&`.

LK4D4 commented on api/server/router/local/image.go:
```diff
@@ -284,6 +294,18 @@ func (s *router) postBuild(ctx context.Context, w http.ResponseWriter, r *http.R
+			return err
+		}
+		w.Write(sf.FormatError(errors.New(utils.GetErrorMessage(err))))
+		return nil
```
Hmm, so, we're not returning errors now from this function at all?

tiborvass replied:
i just replicated the behavior from the existing code. See https://github.com/docker/docker/pull/16147/files#diff-3e7b06b11f8860a92b1bcb4a17ac332dL351

LK4D4 replied:
Yeah, but before many errors was returned from this func, for example from `json.NewEncoder`, not it'll be always nil. If I understand correctly that was only for `Build` func, which you replaced with `b.Build`.

tiborvass replied:
@LK4D4 think of it this way: the logic in `errf` is always correct: only send the error code back from the handler if the HTTP header wasn't already written (the output.Flushed() condition). If it was already sent, then the error doesn't matter at all that's why it's nil. In that case we have to write the error in the body. So technically the json encode error should have the same behavior as the previous code.

LK4D4 replied:
Why this func return errors at all? :)

tiborvass replied:
@LK4D4 for the case when the HTTP header wasn't written yet, it returns the error, which is then translated by httputil.WriteError (called from api/server/server.go) to an HTTP status code.

LK4D4 commented on api/server/router/local/image.go:
```diff
@@ -284,6 +294,18 @@ func (s *router) postBuild(ctx context.Context, w http.ResponseWriter, r *http.R
+		if !output.Flushed() {
+			return err
+		}
+		w.Write(sf.FormatError(errors.New(utils.GetErrorMessage(err))))
```
I think Write error deserves writing to log. And why it's `w` and not `output` here?

tiborvass replied:
ok, but for the record: it would be an addition to the existing behavior

LK4D4 commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+// NewBuilder creates a new Dockerfile builder from an optional dockerfile and a Config.
+// If dockerfile is nil, the Dockerfile specified by Config.DockerfileName,
+// will be read from the Context passed to Build().
+func NewBuilder(dockerfile io.ReadCloser, config *Config) (b *Builder, err error) {
```
Looks like first argument always is nil.

tiborvass replied:
Won't be the case for client side builder. This is so that we can provide a dockerfile independent from the context. It will get the Dockerfile from the context only if this dockerfile is nil. Which is always the case for the server side implementation.

LK4D4 commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+// * NOT tag the image, that is responsibility of the caller.
+//
+func (b *Builder) Build(docker builder.Docker, context builder.Context) (builder.ImageID, error) {
+	b.docker = docker
```
this looks weird as for me, why not setup them in `NewBuilder`? We create it every time anyway.

LK4D4 commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+
+	// check if there are any leftover build-args that were passed but not
+	// consumed during build. Return an error, if there are any.
+	leftoverArgs := []string{}
```
looks like all other code uses `var slice []...`

LK4D4 commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+// - call parse.Parse() to get AST root from Dockerfile entries
+// - do build by calling builder.dispatch() to call all entries' handling routines
+// TODO: remove?
+func BuildFromConfig(d *daemon.Daemon, config *runconfig.Config, changes []string) (*runconfig.Config, error) {
```
Looks like daemon is unused here and that looks weird :)

LK4D4 commented on builder/dockerfile/builder.go:
```diff
@@ -0,0 +1,291 @@
+		}
+	}
+
+	b, _ := NewBuilder(nil, nil)
```
wow, you shouldn't skip errors only because you know that it won't happen.

LK4D4 commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,26 +209,27 @@ func from(b *builder, args []string, attributes map[string]bool, original string
-		// note that the top level err will still be !nil here if IsNotExist is
-		// not the error. This approach just simplifies the logic a bit.
+	if shouldPull {
+		image, err = b.docker.Pull(name)
```
Just do this in `image == nil` block and get rid of `shouldPull` variable.

LK4D4 commented on builder/dockerfile/dispatchers.go:
```diff
@@ -208,26 +209,27 @@ func from(b *builder, args []string, attributes map[string]bool, original string
 	}
-
 	return b.processImageFrom(image)
+
```
I don't like it :D

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+	attrs := ast.Attributes
+	original := ast.Original
+	flags := ast.Flags
+	strs := []string{}
```
should be defined before loop as `var`.

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+		return err
+	}
+
+	attrs := ast.Attributes
```
Used only once, just use `ast.Attributes`.

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	attrs := ast.Attributes
+	original := ast.Original
```
same

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+
+	attrs := ast.Attributes
+	original := ast.Original
+	flags := ast.Flags
```
used as `ast.Flags` below :D

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+	copy(strList, strs)
+	msgList := make([]string, n)
+
+	var i int
```
looks like good candidate for initial value of `for ast.Next != nil` loop.

LK4D4 commented on builder/dockerfile/evaluator.go:
```diff
@@ -0,0 +1,193 @@
+		}
+		envs = append(envs, fmt.Sprintf("%s=%s", key, val))
+	}
+	for ast.Next != nil {
```
Yeah, I was wrong about vars before, but maybe it's like better to save `ast` untouched and use another variable for this loop.

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+	parts := strings.Split(path, string(os.PathSeparator))
+	filename := parts[len(parts)-1]
+	if filename == "" {
+		err = fmt.Errorf("cannot determine filename from url: %s", u)
```
just return this error here.

tiborvass replied:
@LK4D4 i'm using named return values

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+		tmpFile.Close()
+		return
+	}
+	fmt.Fprintln(b.Stdout)
```
Maybe deserves comment, because I'm not sure what is it.

tiborvass replied:
@LK4D4 me neither: https://github.com/docker/docker/pull/16147/files#diff-2aef7cf7f54fe31bf022d583b6ca4ca6L339

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+	}
+	fmt.Fprintln(b.Stdout)
+	// ignoring error because the file was already opened successfully
+	tmpFileSt, _ := tmpFile.Stat()
```
I think successful open is not guarantee of successful stat. That's why `Stat` method returns error.

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+				return err
+			}
+			if info.Name() == "" {
+				// Why are we doing this check?
```
dunno, lol

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+	if err != nil {
+		return nil, err
+	}
+	copyInfos := []copyInfo{{FileInfo: fi, decompress: allowLocalDecompression}}
```
maybe break lines, hard to read

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+
+	// Copy the ONBUILD triggers, and remove them from the config, since the config will be committed.
+	onBuildTriggers := b.runConfig.OnBuild
+	b.runConfig.OnBuild = []string{}
```
Not sure, but maybe nil?

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+		c.Path = s[0]
+		c.Args = s[1:]
+	} else {
+		config.Cmd = stringutils.NewStrSlice()
```
I'm not sure what this mean, config is a copy and it doesn't seem to left this func.

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+	}
+	if f, ok := f.(*os.File); ok {
+		// ignoring error because Open already succeeded
+		fi, _ := f.Stat()
```
same as before, of course you opened if you have `os.File`, but `Stat` method still returns error :)

LK4D4 commented on builder/dockerfile/internals.go:
```diff
@@ -0,0 +1,657 @@
+// determine if build arg is part of built-in args or user
+// defined args in Dockerfile at any point in time.
+func (b *Builder) isBuildArgAllowed(arg string) bool {
+	if _, ok := BuiltinAllowedBuildArgs[arg]; ok {
```
if you use `map[string]bool` for this maps you can write nice logical expression here

LK4D4 commented on builder/dockerfile/support.go:
```diff
@@ -0,0 +1,16 @@
+
+func handleJSONArgs(args []string, attributes map[string]bool) []string {
+	if len(args) == 0 {
+		return []string{}
```
why not args or nil? :)

LK4D4 commented on builder/dockerignore.go:
```diff
@@ -0,0 +1,46 @@
+	excludes, _ := utils.ReadDockerIgnore(dockerignore)
+	filesToRemove = append([]string{".dockerignore"}, filesToRemove...)
+	for _, fileToRemove := range filesToRemove {
+		if rm, _ := fileutils.Matches(fileToRemove, excludes); rm {
```
I think in two lines will be cleaner.

LK4D4 commented on builder/remote.go:
```diff
@@ -0,0 +1,119 @@
+			// use fallback handler supplied by caller (for raw tar contexts)
+			contentType = ""
+			// using goto because the defer-close in the 'if' above would make the code more complicated to read.
+			goto handleContentType
```
I didn't understand this. Seems like you just need always close `contextReader`. I'm pretty sure that without goto will be simpler.

LK4D4 commented on builder/tarsum.go:
```diff
@@ -0,0 +1,166 @@
+	cleanpath = filepath.Clean(string(os.PathSeparator) + path)[1:]
+	fullpath, err = symlink.FollowSymlinkInScope(filepath.Join(c.root, path), c.root)
+	if err != nil {
+		err = fmt.Errorf("Forbidden path outside the build context: %s (%s)", path, fullpath)
```
I think you should return full arguments in case of error. Maybe invalidate cleanpath and fullpath.

tiborvass replied:
@LK4D4 what do you mean by full arguments?

LK4D4 replied:
I mean like `return "", "", fmt.Errrof("")`

LK4D4 commented on daemon/daemonbuilder/builder.go:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	if src.IsDir() {
+
```
too much spaces

LK4D4 commented on daemon/daemonbuilder/builder.go:
```diff
@@ -0,0 +1,239 @@
+		}
+		return fixPermissions(srcPath, destPath, 0, 0, destExists)
+
+	} else if decompress {
```
don't need else here

crosbymichael commented:
LGTM

tiborvass commented:
Squashed commits

LK4D4 commented:
LGTM

jfrazelle commented:
LGTM

cpuguy83 commented:
LGTM
Tests are passing, merging this puppy.

vdemeester commented:
:tada:
